File: Go/source/idgen/SnowWorkerM1.go
/*
 * 版权属于：yitter(yitter@126.com)
 * 代码编辑：guoyahao
 * 代码修订：yitter
 * 开源地址：https://github.com/yitter/idgenerator
 */

package idgen

import (
	"sync"
	"time"
)

// SnowWorkerM1 .
type SnowWorkerM1 struct {
	BaseTime          int64  // 基础时间
	WorkerId          uint16 // 机器码
	WorkerIdBitLength byte   // 机器码位长
	SeqBitLength      byte   // 自增序列数位长
	MaxSeqNumber      uint32 // 最大序列数（含）
	MinSeqNumber      uint32 // 最小序列数（含）
	TopOverCostCount  uint32 // 最大漂移次数
	_TimestampShift   byte
	_CurrentSeqNumber uint32

	_LastTimeTick           int64
	_TurnBackTimeTick       int64
	_TurnBackIndex          byte
	_IsOverCost             bool
	_OverCostCountInOneTerm uint32
	// _GenCountInOneTerm      uint32
	// _TermIndex              uint32

	sync.Mutex
}

// NewSnowWorkerM1 .
func NewSnowWorkerM1(options *IdGeneratorOptions) ISnowWorker {
	var workerIdBitLength byte
	var seqBitLength byte
	var maxSeqNumber uint32

	// 1.BaseTime
	var baseTime int64
	if options.BaseTime != 0 {
		baseTime = options.BaseTime
	} else {
		baseTime = 1582136402000
	}

	// 2.WorkerIdBitLength
	if options.WorkerIdBitLength == 0 {
		workerIdBitLength = 6
	} else {
		workerIdBitLength = options.WorkerIdBitLength
	}

	// 3.WorkerId
	var workerId = options.WorkerId

	// 4.SeqBitLength
	if options.SeqBitLength == 0 {
		seqBitLength = 6
	} else {
		seqBitLength = options.SeqBitLength
	}

	// 5.MaxSeqNumber
	if options.MaxSeqNumber <= 0 {
		maxSeqNumber = (1 << seqBitLength) - 1
	} else {
		maxSeqNumber = options.MaxSeqNumber
	}

	// 6.MinSeqNumber
	var minSeqNumber = options.MinSeqNumber

	// 7.TopOverCostCount
	var topOverCostCount = options.TopOverCostCount
	// if topOverCostCount == 0 {
	// 	topOverCostCount = 2000
	// }

	// 8.Others
	timestampShift := (byte)(workerIdBitLength + seqBitLength)
	currentSeqNumber := minSeqNumber

	return &SnowWorkerM1{
		BaseTime:          baseTime,
		WorkerIdBitLength: workerIdBitLength,
		WorkerId:          workerId,
		SeqBitLength:      seqBitLength,
		MaxSeqNumber:      maxSeqNumber,
		MinSeqNumber:      minSeqNumber,
		TopOverCostCount:  topOverCostCount,
		_TimestampShift:   timestampShift,
		_CurrentSeqNumber: currentSeqNumber,

		_LastTimeTick:           0,
		_TurnBackTimeTick:       0,
		_TurnBackIndex:          0,
		_IsOverCost:             false,
		_OverCostCountInOneTerm: 0,
		// _GenCountInOneTerm:      0,
		// _TermIndex:              0,
	}
}

// DoGenIDAction .
func (m1 *SnowWorkerM1) DoGenIdAction(arg *OverCostActionArg) {

}

func (m1 *SnowWorkerM1) BeginOverCostAction(useTimeTick int64) {

}

func (m1 *SnowWorkerM1) EndOverCostAction(useTimeTick int64) {
	// if m1._TermIndex > 10000 {
	// 	m1._TermIndex = 0
	// }
}

func (m1 *SnowWorkerM1) BeginTurnBackAction(useTimeTick int64) {

}

func (m1 *SnowWorkerM1) EndTurnBackAction(useTimeTick int64) {

}

func (m1 *SnowWorkerM1) NextOverCostId() int64 {
	currentTimeTick := m1.GetCurrentTimeTick()
	if currentTimeTick > m1._LastTimeTick {
		// m1.EndOverCostAction(currentTimeTick)
		m1._LastTimeTick = currentTimeTick
		m1._CurrentSeqNumber = m1.MinSeqNumber
		m1._IsOverCost = false
		m1._OverCostCountInOneTerm = 0
		// m1._GenCountInOneTerm = 0
		return m1.CalcId(m1._LastTimeTick)
	}
	if m1._OverCostCountInOneTerm >= m1.TopOverCostCount {
		// m1.EndOverCostAction(currentTimeTick)
		m1._LastTimeTick = m1.GetNextTimeTick()
		m1._CurrentSeqNumber = m1.MinSeqNumber
		m1._IsOverCost = false
		m1._OverCostCountInOneTerm = 0
		// m1._GenCountInOneTerm = 0
		return m1.CalcId(m1._LastTimeTick)
	}
	if m1._CurrentSeqNumber > m1.MaxSeqNumber {
		m1._LastTimeTick++
		m1._CurrentSeqNumber = m1.MinSeqNumber
		m1._IsOverCost = true
		m1._OverCostCountInOneTerm++
		// m1._GenCountInOneTerm++

		return m1.CalcId(m1._LastTimeTick)
	}

	// m1._GenCountInOneTerm++
	return m1.CalcId(m1._LastTimeTick)
}

// NextNormalID .
func (m1 *SnowWorkerM1) NextNormalId() int64 {
	currentTimeTick := m1.GetCurrentTimeTick()
	if currentTimeTick < m1._LastTimeTick {
		if m1._TurnBackTimeTick < 1 {
			m1._TurnBackTimeTick = m1._LastTimeTick - 1
			m1.BeginTurnBackAction(m1._TurnBackTimeTick)
		}

		m1._TurnBackIndex++
		// 每毫秒序列数的前5位是预留位，0用于手工新值，1-4是时间回拨次序
		// 支持4次回拨次序（避免回拨重叠导致ID重复），可无限次回拨（次序循环使用）。
		if m1._TurnBackIndex > 4 {
			m1._TurnBackIndex = 1
		}

		// time.Sleep(time.Duration(1) * time.Millisecond)
		return m1.CalcTurnBackId(m1._TurnBackTimeTick)
	}

	// 时间追平时，_TurnBackTimeTick清零
	if m1._TurnBackTimeTick > 0 {
		m1.EndTurnBackAction(m1._TurnBackTimeTick)
		m1._TurnBackTimeTick = 0
	}

	if currentTimeTick > m1._LastTimeTick {
		m1._LastTimeTick = currentTimeTick
		m1._CurrentSeqNumber = m1.MinSeqNumber
		return m1.CalcId(m1._LastTimeTick)
	}

	if m1._CurrentSeqNumber > m1.MaxSeqNumber {
		m1.BeginOverCostAction(currentTimeTick)
		// m1._TermIndex++
		m1._LastTimeTick++
		m1._CurrentSeqNumber = m1.MinSeqNumber
		m1._IsOverCost = true
		m1._OverCostCountInOneTerm = 1
		// m1._GenCountInOneTerm = 1

		return m1.CalcId(m1._LastTimeTick)
	}

	return m1.CalcId(m1._LastTimeTick)
}

// CalcID .
func (m1 *SnowWorkerM1) CalcId(useTimeTick int64) int64 {
	result := int64(useTimeTick<<m1._TimestampShift) + int64(m1.WorkerId<<m1.SeqBitLength) + int64(m1._CurrentSeqNumber)
	m1._CurrentSeqNumber++
	return result
}

// CalcTurnBackID .
func (m1 *SnowWorkerM1) CalcTurnBackId(useTimeTick int64) int64 {
	result := int64(useTimeTick<<m1._TimestampShift) + int64(m1.WorkerId<<m1.SeqBitLength) + int64(m1._TurnBackIndex)
	m1._TurnBackTimeTick--
	return result
}

// GetCurrentTimeTick .
func (m1 *SnowWorkerM1) GetCurrentTimeTick() int64 {
	var millis = time.Now().UnixNano() / 1e6
	return millis - m1.BaseTime
}

// GetNextTimeTick .
func (m1 *SnowWorkerM1) GetNextTimeTick() int64 {
	tempTimeTicker := m1.GetCurrentTimeTick()
	for tempTimeTicker <= m1._LastTimeTick {
		time.Sleep(time.Duration(1) * time.Millisecond)
		tempTimeTicker = m1.GetCurrentTimeTick()
	}
	return tempTimeTicker
}

// NextId .
func (m1 *SnowWorkerM1) NextId() int64 {
	m1.Lock()
	defer m1.Unlock()
	if m1._IsOverCost {
		return m1.NextOverCostId()
	} else {
		return m1.NextNormalId()
	}
}

File: Rust/source/src/idgen/snow_worker_m1.rs
/*
 * 版权属于：yitter(yitter@126.com)
 * 开源地址：https://github.com/yitter/idgenerator
 */
use crate::idgen::*;
use chrono::Utc;
use std::thread;
use std::thread::sleep;
// use lazy_static::lazy_static;

pub struct SnowWorkerM1 {
    ///基础时间
    pub BaseTime: i64,
    ///机器码
    pub WorkerId: u32,
    ///机器码位长
    pub WorkerIdBitLength: u8,
    ///自增序列数位长
    pub SeqBitLength: u8,
    ///最大序列数（含）
    pub MaxSeqNumber: u32,
    ///最小序列数（含）
    pub MinSeqNumber: u32,
    ///最大漂移次数
    pub TopOverCostCount: u32,

    _TimestampShift: u8,
    _CurrentSeqNumber: u32,
    _LastTimeTick: i64,
    _TurnBackTimeTick: i64,
    _TurnBackIndex: u8,
    _IsOverCost: bool,
    _OverCostCountInOneTerm: u32,
    _GenCountInOneTerm: u32,
    _TermIndex: u32,
}

impl SnowWorkerM1 {
    pub fn Default() -> SnowWorkerM1 {
        let options = IdGeneratorOptions::New(1);
        return SnowWorkerM1::New(options);
    }

    pub fn SetOptions(&mut self, options: IdGeneratorOptions) {
        // 1.BaseTime
        if options.BaseTime == 0 {
            self.BaseTime = 1582136402000;
        } else if options.BaseTime < 631123200000
            || options.BaseTime > Utc::now().timestamp_millis()
        {
            panic!("BaseTime error.");
        } else {
            self.BaseTime = options.BaseTime;
        }

        // 2.WorkerIdBitLength
        if options.WorkerIdBitLength <= 0 {
            panic!("WorkerIdBitLength error.(range:[1, 21])");
        }
        if options.SeqBitLength + options.WorkerIdBitLength > 22 {
            panic!("error：WorkerIdBitLength + SeqBitLength <= 22");
        } else {
            // self.WorkerIdBitLength = options.WorkerIdBitLength;
            self.WorkerIdBitLength = if options.WorkerIdBitLength <= 0 {
                6
            } else {
                options.WorkerIdBitLength
            };
        }

        // 3.WorkerId
        let mut maxWorkerIdNumber = (1 << options.WorkerIdBitLength) - 1;
        if maxWorkerIdNumber == 0 {
            maxWorkerIdNumber = 63;
        }
        if options.WorkerId < 0 || options.WorkerId > maxWorkerIdNumber {
            panic!("WorkerId error. (range:[0, {} ]", maxWorkerIdNumber);
        } else {
            self.WorkerId = options.WorkerId;
        }

        // 4.SeqBitLength
        if options.SeqBitLength < 2 || options.SeqBitLength > 21 {
            panic!("SeqBitLength error. (range:[2, 21])");
        } else {
            // self.SeqBitLength = options.SeqBitLength;
            self.SeqBitLength = if options.SeqBitLength <= 0 {
                6
            } else {
                options.SeqBitLength
            };
        }

        // 5.MaxSeqNumber
        let mut maxSeqNumber = (1 << options.SeqBitLength) - 1;
        if maxSeqNumber == 0 {
            maxSeqNumber = 63;
        }
        if options.MaxSeqNumber < 0 || options.MaxSeqNumber > maxSeqNumber {
            panic!("MaxSeqNumber error. (range:[1, {}]", maxSeqNumber);
        } else {
            self.MaxSeqNumber = if options.MaxSeqNumber == 0 {
                maxSeqNumber
            } else {
                options.MaxSeqNumber
            };
        }

        // 6.MinSeqNumber
        if options.MinSeqNumber < 5 || options.MinSeqNumber > maxSeqNumber {
            panic!("MinSeqNumber error. (range:[5, {}]", maxSeqNumber);
        } else {
            self.MinSeqNumber = options.MinSeqNumber;
            // self.MinSeqNumber = if options.MinSeqNumber <= 0 { 5 } else { options.MinSeqNumber };
        }

        // 7.TopOverCostCount
        //self.TopOverCostCount = if options.TopOverCostCount == 0 { 2000 } else { options.TopOverCostCount };
        if options.TopOverCostCount < 0 || options.TopOverCostCount > 10000 {
            panic!("TopOverCostCount error. (range:[0, 10000]");
        } else {
            self.TopOverCostCount = options.TopOverCostCount;
        }

        // 8.Others
        self._TimestampShift = self.WorkerIdBitLength + self.SeqBitLength;
        self._CurrentSeqNumber = self.MinSeqNumber;

        if options.Method == 1 {
            sleep(std::time::Duration::from_millis(500));
        }
    }

    pub fn New(options: IdGeneratorOptions) -> SnowWorkerM1 {
        let mut worker = SnowWorkerM1 {
            BaseTime: 1582136402000,
            WorkerIdBitLength: 0,
            WorkerId: 0,
            SeqBitLength: 0,
            MaxSeqNumber: 0,
            MinSeqNumber: 0,
            TopOverCostCount: 0,
            _TimestampShift: 0,
            _CurrentSeqNumber: 0,

            _LastTimeTick: 0,
            _TurnBackTimeTick: 0,
            _TurnBackIndex: 0,
            _IsOverCost: false,
            _OverCostCountInOneTerm: 0,
            _GenCountInOneTerm: 0,
            _TermIndex: 0,
        };

        worker.SetOptions(options);
        return worker;
    }

    pub fn NextId(&mut self) -> i64 {
        // println!("SeqBitLength: {}", self.SeqBitLength);
        if self._IsOverCost {
            self.NextOverCostId()
        } else {
            self.NextNormalId()
        }
    }

    fn DoGenIdAction(&self, arg: OverCostActionArg) {}

    fn BeginOverCostAction(&self, useTimeTick: i64) {}

    fn EndOverCostAction(&mut self, useTimeTick: i64) {
        if self._TermIndex > 10000 {
            self._TermIndex = 0;
        }
    }

    fn BeginTurnBackAction(&self, useTimeTick: i64) {}

    fn EndTurnBackAction(&self, useTimeTick: i64) {}

    fn NextOverCostId(&mut self) -> i64 {
        let currentTimeTick = self.GetCurrentTimeTick();

        if currentTimeTick > self._LastTimeTick {
            self.EndOverCostAction(currentTimeTick);

            self._LastTimeTick = currentTimeTick;
            self._CurrentSeqNumber = self.MinSeqNumber;
            self._IsOverCost = false;
            self._OverCostCountInOneTerm = 0;
            self._GenCountInOneTerm = 0;

            return self.CalcId(self._LastTimeTick);
        }

        if self._OverCostCountInOneTerm >= self.TopOverCostCount {
            self.EndOverCostAction(currentTimeTick);

            self._LastTimeTick = self.GetNextTimeTick();
            self._CurrentSeqNumber = self.MinSeqNumber;
            self._IsOverCost = false;
            self._OverCostCountInOneTerm = 0;
            self._GenCountInOneTerm = 0;

            return self.CalcId(self._LastTimeTick);
        }

        if self._CurrentSeqNumber > self.MaxSeqNumber {
            self._LastTimeTick += 1;
            self._CurrentSeqNumber = self.MinSeqNumber;
            self._IsOverCost = true;
            self._OverCostCountInOneTerm += 1;
            self._GenCountInOneTerm += 1;

            return self.CalcId(self._LastTimeTick);
        }

        self._GenCountInOneTerm += 1;
        return self.CalcId(self._LastTimeTick);
    }

    fn NextNormalId(&mut self) -> i64 {
        let currentTimeTick = self.GetCurrentTimeTick();

        if currentTimeTick < self._LastTimeTick {
            if self._TurnBackTimeTick < 1 {
                self._TurnBackTimeTick = self._LastTimeTick - 1;
                self.BeginTurnBackAction(self._TurnBackTimeTick);
            }

            self._TurnBackIndex += 1;
            // 每毫秒序列数的前5位是预留位，0用于手工新值，1-4是时间回拨次序
            // 支持4次回拨次序（避免回拨重叠导致ID重复），可无限次回拨（次序循环使用）。
            if self._TurnBackIndex > 4 {
                self._TurnBackIndex = 1;
            }

            // thread::sleep(std::time::Duration::from_millis(1));
            return self.CalcTurnBackId(self._TurnBackTimeTick);
        }

        // 时间追平时，_TurnBackTimeTick清零
        if self._TurnBackTimeTick > 0 {
            self.EndTurnBackAction(self._TurnBackTimeTick);
            self._TurnBackTimeTick = 0;
        }

        if currentTimeTick > self._LastTimeTick {
            self._LastTimeTick = currentTimeTick;
            self._CurrentSeqNumber = self.MinSeqNumber;

            return self.CalcId(self._LastTimeTick);
        }

        if self._CurrentSeqNumber > self.MaxSeqNumber {
            self.BeginOverCostAction(currentTimeTick);

            self._TermIndex += 1;
            self._LastTimeTick += 1;
            self._CurrentSeqNumber = self.MinSeqNumber;
            self._IsOverCost = true;
            self._OverCostCountInOneTerm = 1;
            self._GenCountInOneTerm = 1;

            return self.CalcId(self._LastTimeTick);
        }

        return self.CalcId(self._LastTimeTick);
    }

    fn CalcId(&mut self, useTimeTick: i64) -> i64 {
        let result = (useTimeTick << self._TimestampShift)
            + (self.WorkerId << self.SeqBitLength) as i64
            + (self._CurrentSeqNumber) as i64;
        self._CurrentSeqNumber += 1;
        return result;
    }

    fn CalcTurnBackId(&mut self, useTimeTick: i64) -> i64 {
        let result = (useTimeTick << self._TimestampShift)
            + (self.WorkerId << self.SeqBitLength) as i64
            + (self._TurnBackIndex) as i64;
        self._TurnBackTimeTick -= 1;
        return result;
    }

    fn GetCurrentTimeTick(&self) -> i64 {
        return Utc::now().timestamp_millis() - self.BaseTime;
    }

    fn GetNextTimeTick(&self) -> i64 {
        let mut tempTimeTicker = self.GetCurrentTimeTick();

        while tempTimeTicker <= self._LastTimeTick {
            // 暂停1ms
            sleep(std::time::Duration::from_millis(1));
            tempTimeTicker = self.GetCurrentTimeTick();
        }

        return tempTimeTicker;
    }
}
